README.md
# Firebrew

Firefox add-ons manager for CUI.

## Installation

Add this line to your application's Gemfile:

    gem 'firebrew'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install firebrew


_NOTE: This program execution requires the OpenSSL._

## Usage

The structure of the command line is shown below:

```bash
$ firebrew [--help] [--version]
           [--base-dir=<path>] [--profile=<name>] [--firefox=<path>]
           <command> [<args>]
```

### commands

#### install

Install the extension which is designated by the `extension-name` argument:

```bash
$ firebrew install <extension-name>
```

#### uninstall

Uninstall the extension which is designated by the `extension-name` argument:

```bash
$ firebrew uninstall <extension-name>
```

#### info

Show detail information of the extension which is designated by the `extension-name` argument:

```bash
$ firebrew info <extension-name>
```

#### search

Enumerate the remote extensions whose name is matched the `term` argument:

```bash
$ firebrew search <term>
```

#### list

Enumerate the installed extensions:

```bash
$ firebrew list
```

#### profile

Show the profile information:

```bash
$ firebrew profile [--attribute=<attr-name>]
```

##### options

###### --attribute

The name of the attribute which want to display:

```bash
-a <attr-name>, --attribute=<attr-name>
```

### global options

#### --base-dir

The Firefox profiles.ini directory:

```bash
-d <path>, --base-dir=<path>
```

The default value is listed below:

| platform | value |
| -------- | ----- |
| Mac OS X | `~/Library/Application Support/Firefox` | 
| Linux    | `~/.mozilla/firefox` |
| Windows  | `%APPDATA%\Mozilla\Firefox` |

It's able to overridden by the `FIREBREW_FIREFOX_PROFILE_BASE_DIR` environment variable.

#### --profile-name

The Firefox profile name:

```bash
-p <name>, --profile=<name>
```

The default value is `default`, and it's able to overridden by the `FIREBREW_FIREFOX_PROFILE` environment variable.

#### --firefox

The Firefox command path:

```bash
-f <path>, --firefox=<path>
```

The default value is listed below:

| platform | value |
| -------- | ----- |
| Mac OS X | `/Applications/Firefox.app/Contents/MacOS/firefox-bin` |
| Linux    | `/usr/bin/firefox` |
| Windows  | `%PROGRAMFILES%\Mozilla Firefox\firefox.exe` or `%PROGRAMFILES(X86)%\Mozilla Firefox\firefox.exe` |

It's able to overridden by the `FIREBREW_FIREFOX` environment variable.


## Contributing

1. Fork it ( http://github.com/mrk21/firebrew/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request



CHANGELOG.md
# Changelog

## [0.3.0](https://github.com/mrk21/firebrew/tree/v0.3.0) - 2014-09-30

* Future [#36](https://github.com/mrk21/firebrew/issues/36): Implement the `os` gem
* Future [#18](https://github.com/mrk21/firebrew/issues/18): Make it possible to display a download progress of the extension

## [0.2.0](https://github.com/mrk21/firebrew/tree/v0.2.0) - 2014-08-31

* Future [#35](https://github.com/mrk21/firebrew/issues/35): Add the subcommand help
* Future [#34](https://github.com/mrk21/firebrew/issues/34): Make the help message in detail
* Future [#33](https://github.com/mrk21/firebrew/issues/33): Change to 0 a return value of the help command and version command
* Bugfix [#32](https://github.com/mrk21/firebrew/issues/32): The `OptionParser` exceptions has not been handled enough
* Bugfix [#31](https://github.com/mrk21/firebrew/issues/31): Has not been throw the `SystemCall` exception when executed the command which is not existed
* Future [#30](https://github.com/mrk21/firebrew/issues/30): Add the detailed error messages
* Future [#29](https://github.com/mrk21/firebrew/issues/29): Change the command return value
* Future [#28](https://github.com/mrk21/firebrew/issues/28): Stop depending on `ActiveSupport`
* Future [#27](https://github.com/mrk21/firebrew/issues/27): Stop depending on `ActiveResource`
* Future [#26](https://github.com/mrk21/firebrew/issues/26): Stop depending on `ActiveModel`
* Future [#25](https://github.com/mrk21/firebrew/issues/25): Add the command of getting the profile information
* Future [#16](https://github.com/mrk21/firebrew/issues/16): The handling of when the network errors of the 404 Not Found HTTP error, etc occurs on the Amo::Search

## [0.1.3](https://github.com/mrk21/firebrew/tree/v0.1.3) - 2014-08-19

* Bugfix [#24](https://github.com/mrk21/firebrew/issues/24): When the `em:unpack` value of the install manifests was true, is unable to normally installing

## [0.1.2](https://github.com/mrk21/firebrew/tree/v0.1.2) - 2014-08-15

* Bugfix [#23](https://github.com/mrk21/firebrew/issues/23): If the `install` element of the extension which was got by the generic AMO API was equal or greater than two, then occurs errors
* Future [#22](https://github.com/mrk21/firebrew/issues/22): Designate the current platform to the `os` parameter when use the `Amo::Search`

## [0.1.1](https://github.com/mrk21/firebrew/tree/v0.1.1) - 2014-07-16

* Bugfix [#12](https://github.com/mrk21/firebrew/issues/12): The unit test failed on the Windows(#10): The 2 to 5 errors
* Bugfix [#11](https://github.com/mrk21/firebrew/issues/11): The unit test failed on the Windows(#10): The first error
* Bugfix [#9](https://github.com/mrk21/firebrew/issues/9): Is not able to install on the profile creation immediate aftermath
* Bugfix [#8](https://github.com/mrk21/firebrew/issues/8): The typo of the `Runner::default_options` for the Windows


lib/firebrew/command_line.rb
require 'optparse'
require 'erb'
require 'firebrew/runner'

module Firebrew
  class CommandLine
    attr_reader :arguments
    
    def self.execute
      begin
        if block_given? then
          yield
        else
          self.new(ARGV).execute
        end
      rescue Firebrew::Error => e
        $stderr.puts e.message
        exit e.status
      rescue SystemExit => e
        exit 0
      rescue Exception => e
        $stderr.puts e.inspect
        $stderr.puts e.backtrace
        exit 1
      else
        exit 0
      end
    end
    
    def initialize(args=[])
      @arguments = {
        command: nil,
        params: {},
        config: {}
      }
      global_parser = self.option_parser do |parser|
        parser.banner = self.desc(<<-DESC)
          Usage: firebrew [--help] [--version]
                 [--base-dir=<path>] [--profile=<name>] [--firefox=<path>]
                 <command> [<args>]
        DESC
        
        parser.separator ''
        parser.separator 'commands:'
        begin
          pos = 11
          
          self.summary(parser, :install, <<-DESC, pos)
            Install the Firefox extension
          DESC
          
          self.summary(parser, :uninstall, <<-DESC, pos)
            Uninstall the Firefox extension
          DESC
          
          self.summary(parser, :info, <<-DESC, pos)
            Show detail information of the Firefox extension
          DESC
          
          self.summary(parser, :search, <<-DESC, pos)
            Search Firefox extensions
          DESC
          
          self.summary(parser, :list, <<-DESC, pos)
            Enumerate the installed Firefox extensions
          DESC
          
          self.summary(parser, :profile, <<-DESC, pos)
            Show the profile information
          DESC
        end
      end
      
      global_parser.order!(args)
      command = args.shift.to_s.intern
      
      case command
      when :install, :uninstall, :info then
        subcommand_parser = self.option_parser do |parser|
          parser.banner = self.desc(<<-DESC)
            Usage: firebrew [--help] [--version]
                   [--base-dir=<path>] [--profile=<name>] [--firefox=<path>]
                   #{command} <extension-name>
          DESC
        end
        
        subcommand_parser.permute!(args)
        self.arguments[:command] = command
        self.arguments[:params][:term] = args[0]
        
      when :search then
        subcommand_parser = self.option_parser do |parser|
          parser.banner = self.desc(<<-DESC)
            Usage: firebrew [--help] [--version]
                   [--base-dir=<path>] [--profile=<name>] [--firefox=<path>]
                   #{command} <term>
          DESC
        end
        
        subcommand_parser.permute!(args)
        self.arguments[:command] = command
        self.arguments[:params][:term] = args[0]
        
      when :list then
        subcommand_parser = self.option_parser do |parser|
          parser.banner = self.desc(<<-DESC)
            Usage: firebrew [--help] [--version]
                   [--base-dir=<path>] [--profile=<name>] [--firefox=<path>]
                   #{command}
          DESC
        end
        
        subcommand_parser.permute!(args)
        self.arguments[:command] = command
        
      when :profile then
        subcommand_parser = self.option_parser do |parser|
          parser.summary_width = 30
          parser.banner = self.desc(<<-DESC)
            Usage: firebrew [--help] [--version]
                   [--base-dir=<path>] [--profile=<name>] [--firefox=<path>]
                   #{command} [--attribute=<attr-name>]
          DESC
          
          parser.separator ''
          parser.separator 'options:'
          begin
            chooses = %r[#{Firebrew::Firefox::Profile.attributes.join('|')}]
            parser.on('-a <attr-name>','--attribute=<attr-name>', chooses, self.desc(<<-DESC)) do |v|
              The name of the attribute which want to display
            DESC
              self.arguments[:params][:attribute] = v
            end
          end
        end
        
        subcommand_parser.permute!(args)
        self.arguments[:command] = command
        
      when :'' then
        global_parser.permute(['--help'])
        
      else
        raise Firebrew::CommandLineError, "Invalid command: #{command}"
      end
      
    rescue OptionParser::ParseError => e
      m = e.message
      m[0] = m[0].upcase
      raise Firebrew::CommandLineError, m
    end
    
    def execute
      runner = Runner.new(self.arguments[:config], true)
      
      case self.arguments[:command]
      when :search, :list then
        results = runner.send(self.arguments[:command], self.arguments[:params])
        results.each do |result|
          puts result.name
        end
        
      when :info then
        result = runner.send(self.arguments[:command], self.arguments[:params])
        puts result.data
        
      when :profile then
        r = runner.profile
        attr = self.arguments[:params][:attribute]
        if attr.nil? then
          attrs = r.class.attributes
          puts ERB.new(self.desc(<<-XML),nil,'-').result(binding)
            <profile>
            <% attrs.each do |attr| -%>
              <<%= attr %>><%= r.send(attr) %></<%= attr %>>
            <% end -%>
            </profile>
          XML
        else
          puts r.send(attr)
        end
        
      else
        runner.send(self.arguments[:command], self.arguments[:params])
      end
    end
    
    protected
    
    def desc(str)
      lines = str.split(/\n/)
      indent = lines.map{|v| v.match(/^ +/).to_a[0].to_s.length}.min
      lines.map{|v| v[indent..-1].rstrip}.join("\n")
    end
    
    def summary(parser, name, description, pos)
      result = ' '*100
      result[0] = name.to_s
      result[pos+1] = self.desc(description)
      result = parser.summary_indent + result.rstrip
      parser.separator result
    end
    
    def option_parser
      parser = OptionParser.new
      parser.version = Firebrew::VERSION
      parser.summary_indent = ' '*3
      parser.summary_width = 25
      
      yield parser
      
      parser.separator ''
      parser.separator 'global options:'
      begin
        parser.on('-d <path>','--base-dir=<path>', String, self.desc(<<-DESC)) do |v|
          Firefox profiles.ini directory
        DESC
          self.arguments[:config][:base_dir] = v
        end
        
        parser.on('-p <name>','--profile=<name>', String, self.desc(<<-DESC)) do |v|
          Firefox profile name
        DESC
          self.arguments[:config][:profile] = v
        end
        
        parser.on('-f <path>','--firefox=<path>', String, self.desc(<<-DESC)) do |v|
          Firefox command path
        DESC
          self.arguments[:config][:firefox] = v
        end
      end
      
      parser.separator ''
      begin
        parser.on('-h', '--help', self.desc(<<-DESC)) do
          Show this message
        DESC
          puts parser.help
          exit
        end
        
        parser.on('-v', '--version', self.desc(<<-DESC)) do
          Show version
        DESC
          puts parser.ver
          exit
        end
      end
      
      parser.separator ''
      parser.separator 'return value:'
      begin
        pos = 3
        
        self.summary(parser, '0', <<-DESC, pos)
          Success
        DESC
        
        self.summary(parser, '1', <<-DESC, pos)
          Error
        DESC
        
        self.summary(parser, '2', <<-DESC, pos)
          No operation
        DESC
      end
      
      return parser
    end
  end
end
